File: src/RandomForest.cpp
#include "RandomForest.h"
#include <cstdlib>
#include <ctime>
#include <algorithm>
#include <map>

namespace {
// Helper to get a random float between min and max
float randFloat(float min, float max) {
    return min + static_cast<float>(rand()) / (static_cast<float>(RAND_MAX/(max-min)));
}
}

// Recursively build a simple tree (random feature, random threshold)
TreeNode* buildTree(const std::vector<UserProfile>& data, int depth, int max_depth) {
    if (data.empty() || depth >= max_depth) {
        // Majority class leaf
        int at_risk = 0;
        for (const auto& u : data) at_risk += (u.healthIndicators.count("AtRisk") && u.healthIndicators.at("AtRisk"));
        TreeNode* leaf = new TreeNode();
        leaf->is_leaf = true;
        leaf->prediction = (at_risk > (int)data.size()/2) ? 1 : 0;
        return leaf;
    }
    // Randomly pick a feature
    std::vector<std::string> features = {"HighBloodPressure", "ChestPain", "Dizziness", "Fatigue & Weakness", "Swelling (Edema)", "Pain in Neck/Jaw/Shoulder/Back", "Excessive Sweating", "Chest Discomfort (Activity)", "Snoring/Sleep Apnea", "Anxiety/Feeling of Doom", "Age"};
    std::string feature = features[rand() % features.size()];
    float threshold = 0;
    if (feature == "Age") {
        int min_age = 200, max_age = -1;
        for (const auto& u : data) {
            min_age = std::min(min_age, u.age);
            max_age = std::max(max_age, u.age);
        }
        threshold = randFloat(min_age, max_age);
    }
    // Split data
    std::vector<UserProfile> left, right;
    for (const auto& u : data) {
        if (feature == "Age") {
            if (u.age < threshold) left.push_back(u);
            else right.push_back(u);
        } else {
            if (u.healthIndicators.count(feature) && u.healthIndicators.at(feature)) left.push_back(u);
            else right.push_back(u);
        }
    }
    TreeNode* node = new TreeNode();
    node->feature = feature;
    node->threshold = threshold;
    node->is_leaf = false;
    node->left = buildTree(left, depth+1, max_depth);
    node->right = buildTree(right, depth+1, max_depth);
    return node;
}

void RandomForest::train(const std::vector<UserProfile>& data, int n_trees, int max_depth) {
    srand((unsigned)time(0));
    trees.clear();
    for (int i = 0; i < n_trees; ++i) {
        // Bootstrap sample
        std::vector<UserProfile> sample;
        for (size_t j = 0; j < data.size(); ++j) sample.push_back(data[rand()%data.size()]);
        trees.push_back(buildTree(sample, 0, max_depth));
    }
}

int predictTree(const TreeNode* node, const UserProfile& user) {
    if (!node) return 0;
    if (node->is_leaf) return node->prediction;
    if (node->feature == "Age") {
        if (user.age < node->threshold) return predictTree(node->left, user);
        else return predictTree(node->right, user);
    } else {
        if (user.healthIndicators.count(node->feature) && user.healthIndicators.at(node->feature))
            return predictTree(node->left, user);
        else
            return predictTree(node->right, user);
    }
}

int RandomForest::predict(const UserProfile& user) const {
    int votes = 0;
    for (const auto& t : trees) votes += predictTree(t, user);
    return (votes > (int)trees.size()/2) ? 1 : 0;
}

float RandomForest::predict_proba(const UserProfile& user) const {
    int votes = 0;
    for (const auto& t : trees) votes += predictTree(t, user);
    return (float)votes / trees.size();
}

void freeTree(TreeNode* node) {
    if (!node) return;
    freeTree(node->left);
    freeTree(node->right);
    delete node;
}

RandomForest::~RandomForest() {
    for (auto t : trees) freeTree(t);
} 

File: src/RandomForest.java
package src;

import java.util.*;

public class RandomForest {
    private List<TreeNode> trees;
    private static final Random random = new Random();

    public RandomForest() {
        this.trees = new ArrayList<>();
    }

    private static float randFloat(float min, float max) {
        return min + random.nextFloat() * (max - min);
    }

    private TreeNode buildTree(List<UserProfile> data, int depth, int maxDepth) {
        if (data.isEmpty() || depth >= maxDepth) {
            int atRisk = 0;
            for (UserProfile u : data) {
                if (u.healthIndicators.getOrDefault("stroke", false)) {
                    atRisk++;
                }
            }
            TreeNode leaf = new TreeNode();
            leaf.isLeaf = true;
            leaf.prediction = (atRisk > data.size() * 0.3) ? 1 : 0;
            return leaf;
        }

        String[] features = {
            "Chest Pain", "Shortness of Breath", "Irregular Heartbeat", 
            "Fatigue & Weakness", "Dizziness", "Swelling (Edema)",
            "Pain in Neck/Jaw/Shoulder/Back", "Excessive Sweating",
            "Persistent Cough", "Nausea/Vomiting", "High Blood Pressure",
            "Chest Discomfort (Activity)", "Cold Hands/Feet",
            "Snoring/Sleep Apnea", "Anxiety/Feeling of Doom", "Age"
        };
        String feature = features[random.nextInt(features.length)];
        float threshold = 0;

        if (feature.equals("Age")) {
            int minAge = Integer.MAX_VALUE;
            int maxAge = Integer.MIN_VALUE;
            for (UserProfile u : data) {
                minAge = Math.min(minAge, u.age);
                maxAge = Math.max(maxAge, u.age);
            }
            threshold = randFloat(minAge, maxAge);
        }

        List<UserProfile> left = new ArrayList<>();
        List<UserProfile> right = new ArrayList<>();

        for (UserProfile u : data) {
            if (feature.equals("Age")) {
                if (u.age < threshold) {
                    left.add(u);
                } else {
                    right.add(u);
                }
            } else {
                if (u.healthIndicators.getOrDefault(feature, false)) {
                    left.add(u);
                } else {
                    right.add(u);
                }
            }
        }

        TreeNode node = new TreeNode();
        node.feature = feature;
        node.threshold = threshold;
        node.isLeaf = false;
        node.left = buildTree(left, depth + 1, maxDepth);
        node.right = buildTree(right, depth + 1, maxDepth);
        return node;
    }

    public void train(List<UserProfile> data, int nTrees, int maxDepth) {
        trees.clear();
        for (int i = 0; i < nTrees; i++) {
            List<UserProfile> sample = new ArrayList<>();
            List<UserProfile> atRisk = new ArrayList<>();
            List<UserProfile> notAtRisk = new ArrayList<>();
            
            for (UserProfile u : data) {
                if (u.healthIndicators.getOrDefault("stroke", false)) {
                    atRisk.add(u);
                } else {
                    notAtRisk.add(u);
                }
            }
            
            int sampleSize = Math.min(atRisk.size(), notAtRisk.size());
            for (int j = 0; j < sampleSize; j++) {
                sample.add(atRisk.get(random.nextInt(atRisk.size())));
                sample.add(notAtRisk.get(random.nextInt(notAtRisk.size())));
            }
            
            trees.add(buildTree(sample, 0, maxDepth));
        }
    }

    private int predictTree(TreeNode node, UserProfile user) {
        if (node == null) return 0;
        if (node.isLeaf) return node.prediction;
        if (node.feature.equals("Age")) {
            if (user.age < node.threshold) {
                return predictTree(node.left, user);
            } else {
                return predictTree(node.right, user);
            }
        } else {
            if (user.healthIndicators.getOrDefault(node.feature, false)) {
                return predictTree(node.left, user);
            } else {
                return predictTree(node.right, user);
            }
        }
    }

    public int predict(UserProfile user) {
        int votes = 0;
        for (TreeNode t : trees) {
            votes += predictTree(t, user);
        }
        return (votes > trees.size() * 0.4) ? 1 : 0;
    }

    public float predictProba(UserProfile user) {
        int votes = 0;
        for (TreeNode t : trees) {
            votes += predictTree(t, user);
        }
        return (float) votes / trees.size();
    }
} 